### DISCUSSION TITLE
Add option to password-protect the encryption key

### DISCUSSION THREAD

**Discussion by lomigmegard** [sniptt-official/ots#11] [Ideas]
An interesting optional feature would be to protect the encryption key using a password. The recipient would have to enter the password on the web page, the decryption still fully client-side.

It could be done using derivation directly, stretching the password to derive the encryption key. But that would undermine the encryption security, making the cyphertext as weak/strong as the password. Another approach is to derive a KEK from the password, wrapping the fully random encryption key. This way the cyphertext is not impacted, and the security level of the password is kept client-side.

**Comment by j-l-m** [comment]
Would it be a problem to use a hash of the password as the KEK?

**Comment by lomigmegard** [comment]
@j-l-m In the case of passwords, not all hash functions are good candidates, and a password-based [KDF](https://en.wikipedia.org/wiki/Key_derivation_function) is advised. 

- Possible functions are [Argon2id](https://en.wikipedia.org/wiki/Argon2), [scrypt](https://en.wikipedia.org/wiki/Scrypt) or [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) for more legacy environment. 
- A salt must be used. 
- The cost (iterations/memory if applicable) is to be tweaked depending on the wanted UX.
